### PR TITLE
New __iter__  for Array module

### PR DESCRIPTION
**Pull request by kangzhiq**
#### Brief description of what is fixed or changed
The array iteration is modified to have a new behavior:
Before:
```python
>>> a = Array([[1, 2], [3, 4]])
>>> for i in a:
...     print(i)
...
1
2
3
4
 
```
Now:
```python
>>> a = Array([[1, 2], [3, 4]])
>>> for i in a:
...     print(i)
...
[1, 2]
[3, 4]
```
Besides, the iteration over each item is moved to a new object named `nditer` :
```python
>>> from sympy.tensor.array.nditer import nditer
>>> for i in nditer(a):
...     print(i)
...
1
2
3
2
3
4
```

#### Release Notes

* tensor
  * Modified __iter__ function and added new class named nditer

### CODE REVIEW

File: sympy/tensor/array/nditer.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        from sympy.matrices.matrices import MatrixBase
+
+        try:
+            if isinstance(self._iter, list):
+                result = self._iter[self._idx]
```
Comment (kangzhiq) [inline sympy/tensor/array/nditer.py]: How to dynamically handle the iteration over a multi-dimension list is a great challenge. Especially for the case where an `array` is cast to `list`:
```python
>>> a = Array([[1, 2], [3, 4]])
>>> b = list(a)
>>> b.__class__
<class 'list'>
>>> b[0].__class__
<class 'sympy.tensor.array.dense_ndim_array.ImmutableDenseNDimArray'>
```
I might need some guidance in this problem. :-)

File: sympy/tensor/tests/test_tensor.py [hunk 2]
```diff
@@ -1318,19 +1318,19 @@ def test_valued_tensor_iter():
...
+    #assert list(ba_matrix) == list(BA)
 
     # iteration on VTensMul
     assert list(A(i1)) == [E, px, py, pz]
-    assert list(BA(i1, i2)) == list(ba_matrix)
-    assert list(3 * BA(i1, i2)) == [3 * i for i in list(ba_matrix)]
-    assert list(-5 * BA(i1, i2)) == [-5 * i for i in list(ba_matrix)]
+    #assert list(BA(i1, i2)) == list(ba_matrix)
+    #assert list(3 * BA(i1, i2)) == [3 * i for i in list(ba_matrix)]
+    #assert list(-5 * BA(i1, i2)) == [-5 * i for i in list(ba_matrix)]
```
Comment (kangzhiq) [inline sympy/tensor/tests/test_tensor.py]: New tests might be needed.

File: sympy/tensor/array/nditer.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
+from sympy.tensor.array.dense_ndim_array import DenseNDimArray
+from sympy.tensor.array.sparse_ndim_array import SparseNDimArray
+from sympy.tensor.array.ndim_array import NDimArray
+from sympy.core.compatibility import Iterable
+
+class nditer(object):
```
Comment (Upabjojr) [inline sympy/tensor/array/nditer.py]: SymPy standard is `NDIter`

Comment (Upabjojr) [reply sympy/tensor/array/nditer.py]: This should be a SymPy object, so inherit `Expr`.

File: sympy/tensor/array/nditer.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+    def __init__(self, iterable):
+        from sympy.matrices.matrices import MatrixBase
+
+        if not isinstance(iterable, (Iterable, MatrixBase)):
+            raise NotImplementedError("Data type not yet supported")
+        if isinstance(iterable, list) and isinstance(iterable[0], NDimArray):
```
Comment (Upabjojr) [inline sympy/tensor/array/nditer.py]: why this condition?

File: sympy/tensor/array/nditer.py (hunk 3 above)
Comment (Upabjojr) [reply sympy/tensor/array/nditer.py]: Maybe this object should be called `Flatten`, as what it does it flattening the array.

File: sympy/tensor/array/nditer.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
+from sympy.tensor.array.dense_ndim_array import DenseNDimArray
```
Comment (Upabjojr) [inline sympy/tensor/array/nditer.py]: please no new files, there is a file for array operations.

File: sympy/tensor/array/nditer.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        if isinstance(iterable, list) and isinstance(iterable[0], NDimArray):
+            temp = []
+            for i in range(len(iterable)):
+                temp += iterable[0].tolist()
+            iterable = temp
+        self._iter = iterable
+        self._idx = 0
+
+    def __iter__(self):
+        return self
```
Comment (Upabjojr) [inline sympy/tensor/array/nditer.py]: why not return the iterator here?

File: sympy/tensor/array/nditer.py (hunk 1 above)
Comment (Upabjojr) [reply sympy/tensor/array/nditer.py]: Please ignore `list`. We have `Array` in SymPy. People who use `list` will use 99% of the time `NumPy`, we don't need support for `list`.

Comment (Upabjojr) [reply sympy/tensor/array/nditer.py]: We don't need to redo what NumPy already does... if you want to use NumPy's features, just use NumPy.

File: sympy/tensor/array/nditer.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+    def __next__(self):
+        from sympy.matrices.matrices import MatrixBase
+
+        try:
```
Comment (Upabjojr) [inline sympy/tensor/array/nditer.py]: No `try` please.

File: sympy/tensor/array/nditer.py (hunk 4 above)
Comment (kangzhiq) [reply sympy/tensor/array/nditer.py]: This condition is to handle to case where an array is cast to list by `list()`:
```python
>>> a = list(Array([[0, 1], [2, 3]]))
>>> a.__class__
<class 'list'>
>>> a[0].__class__
<class 'sympy.tensor.array.dense_ndim_array.ImmutableDenseNDimArray'>
```
And the multdimensional `list` is not considered here, as you explained below. I was thinking the same so I only handled this special case of having a mutipledimensional `list` from `list()`.

File: sympy/tensor/array/nditer.py (hunk 6 above)
Comment (kangzhiq) [reply sympy/tensor/array/nditer.py]: I was thinking the `NDIter` itself as an iterator, that is why `self` is returned here.

File: sympy/tensor/array/nditer.py (hunk 1 above)
Comment (kangzhiq) [reply sympy/tensor/array/nditer.py]: But we have tests like:
```
 assert list(sparse_array) == [0, 0, 0, 1]
```
That is why I ran into an error about `list` and tried to handle this special case. However, I have changed the test by using a simple `tolist()` instead of 'list()'.

File: sympy/tensor/array/nditer.py (hunk 7 above)
Comment (kangzhiq) [reply sympy/tensor/array/nditer.py]: I was using `try` to handle the `IndexError` so that the iterator can be stopped properly. I have seen the usage of `try+except` structure in SymPy's codebase. Should it be avoided?

File: sympy/tensor/array/arrayop.py [hunk 8]
```diff
@@ -296,3 +296,57 @@ def permutedims(expr, perm):
         new_array[i] = expr[t]
 
     return type(expr)(new_array, new_shape)
+
+
+class Flatten(object):
```
Comment (Upabjojr) [inline sympy/tensor/array/arrayop.py]: inherit `Basic`

File: sympy/tensor/array/arrayop.py [hunk 9]
```diff
@@ -296,3 +296,57 @@ def permutedims(expr, perm):
...
+
+    def __init__(self, iterable):
+        from sympy.matrices.matrices import MatrixBase
+
+        if not isinstance(iterable, (Iterable, MatrixBase)):
+            raise NotImplementedError("Data type not yet supported")
+        if isinstance(iterable, list) and len(iterable)>0 and isinstance(iterable[0], NDimArray):
```
Comment (Upabjojr) [inline sympy/tensor/array/arrayop.py]: if it's a list, then transform it into an `Array`.

Comment (kangzhiq) [reply sympy/tensor/array/arrayop.py]: If we transform it to `Array`, there will be a loop of calling functions:
While initiating an Array, we will call `Flatten()` in return.

Comment (Upabjojr) [reply sympy/tensor/array/arrayop.py]: You cannot add a `list` to a SymPy object. What about calling `flatten( )` in `Array`?

File: sympy/tensor/array/dense_ndim_array.py [hunk 10]
```diff
@@ -64,13 +64,13 @@ def __getitem__(self, index):
...
-            if isinstance(index, slice):
-                return self._array[index]
-            else:
-                if self.shape == ():
-                    index = ()
-                index = self._parse_index(index)
-                return self._array[index]
+            # if isinstance(index, slice):
+                # return self._array[index]
+            # else:
```
Comment (Upabjojr) [inline sympy/tensor/array/dense_ndim_array.py]: if you are removing code, just remove it

File: sympy/tensor/array/dense_ndim_array.py [hunk 11]
```diff
@@ -148,11 +155,12 @@ def __new__(cls, iterable, shape=None, **kwargs):
     @classmethod
     def _new(cls, iterable, shape, **kwargs):
         from sympy.utilities.iterables import flatten
+        from sympy.tensor.array.arrayop import Flatten
 
         shape, flat_list = cls._handle_ndarray_creation_inputs(iterable, shape, **kwargs)
         shape = Tuple(*map(_sympify, shape))
         cls._check_special_bounds(flat_list, shape)
-        flat_list = flatten(flat_list)
+        flat_list = flatten(Flatten(flat_list))
```
Comment (Upabjojr) [inline sympy/tensor/array/dense_ndim_array.py]: is this really needed? isn't `flatten` already supposed to flatten the list?

File: sympy/tensor/array/arrayop.py (hunk 9 above)
Comment (kangzhiq) [reply sympy/tensor/array/arrayop.py]: OK!

File: sympy/tensor/array/dense_ndim_array.py (hunk 11 above)
Comment (kangzhiq) [reply sympy/tensor/array/dense_ndim_array.py]: No, it can be removed. I didn't think about it. Thank you!!
Since I have modified the `flatten()` function in the last commit, it can now flatten the result of `list(Array())`. 
What we had before is that it would run into the case of having 'args' and transform an `Array` to `Arrays.args`, which will add the `shape` into the flatten list.
```python
 if hasattr(el, 'args'):
    el = el.args
```

**Comment by Upabjojr** [comment]
Commented code needs to be removed.

**Comment by Upabjojr** [comment]
There is a conflict with master. You need to merge the master branch and resolve the conflict.

File: sympy/core/tests/test_args.py [hunk 12]
```diff
@@ -3982,6 +3982,11 @@ def test_sympy__tensor__array__array_comprehension__ArrayComprehension():
     arrcom = ArrayComprehension(x, (x, 1, 5))
     assert _test_args(arrcom)
 
+def test_sympy__tensor__array__arrayop__Flatten():
+    from sympy.tensor.array.arrayop import Flatten
+    from sympy.tensor.array.dense_ndim_array import ImmutableDenseNDimArray
+    fla = Flatten(ImmutableDenseNDimArray(range(24)).reshape(2, 3, 4))
```
Comment (Upabjojr) [inline sympy/core/tests/test_args.py]: where is the call to `_test_args`?

File: sympy/tensor/array/arrayop.py [hunk 13]
```diff
@@ -296,3 +296,73 @@ def permutedims(expr, perm):
...
+
+    Examples
+    ========
+
+    >>> from sympy.tensor.array.arrayop import Flatten
+    >>> from sympy.tensor.array import Array
+    >>> A = Array(range(6)).reshape(2, 3)
+    >>> Flatten(A)
+    Flatten([[0, 1, 2], [3, 4, 5]])
+    >>> [*Flatten(A)]
```
Comment (Upabjojr) [inline sympy/tensor/array/arrayop.py]: is this valid in Python 2.7?

File: sympy/tensor/array/arrayop.py [hunk 14]
```diff
@@ -296,3 +296,73 @@ def permutedims(expr, perm):
...
+        self._iter = iterable
+        self._idx = 0
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        from sympy.matrices.matrices import MatrixBase
+
+        try:
```
Comment (Upabjojr) [inline sympy/tensor/array/arrayop.py]: please remove this `try`. You should define the total number of elements and then count up to them.

File: sympy/tensor/array/ndim_array.py [hunk 15]
```diff
@@ -140,10 +140,7 @@ def _handle_ndarray_creation_inputs(cls, iterable=None, shape=None, **kwargs):
             # Construction of a sparse array from a sparse array
             elif isinstance(iterable, SparseNDimArray):
                 return iterable._shape, iterable._sparse_array
-            # Construction from another `NDimArray`:
-            elif isinstance(iterable, NDimArray):
-                shape = iterable.shape
-                iterable = list(iterable)
```
Comment (Upabjojr) [inline sympy/tensor/array/ndim_array.py]: why was this removed???

Comment (kangzhiq) [reply sympy/tensor/array/ndim_array.py]: Because we are trying to cast an `Array` to `list` and then flatten it in the constructor with `flatten()`. Since `flatten()` can now flatten an `Array` properly, I was thinking that we don't need this case any more.

Comment (Upabjojr) [reply sympy/tensor/array/ndim_array.py]: OK
